### PR TITLE
Update to .NET 6

### PR DESCRIPTION
**Pull request by abcdefg30**
New version of #19236 based on the .NET 6. Also increases the minimum version for macOS packaging to `10.14`.

⚠️ Visual Studio 2019 does not work with .NET 6, you'll need to update to Visual Studio 2022. ⚠️

### CODE REVIEW

File: packaging/functions.sh [hunk 1]
```diff
@@ -153,7 +153,7 @@ install_windows_launcher()
 	rm -rf "${SRC_PATH}/OpenRA.WindowsLauncher/obj"
 	dotnet publish "${SRC_PATH}/OpenRA.WindowsLauncher/OpenRA.WindowsLauncher.csproj" -c Release -r "${TARGETPLATFORM}" -p:LauncherName="${LAUNCHER_NAME}" -p:TargetPlatform="${TARGETPLATFORM}" -p:ModID="${MOD_ID}" -p:DisplayName="${MOD_NAME}" -p:FaqUrl="${FAQ_URL}" -o "${DEST_PATH}" --self-contained true
 
-	# NET 5 is unable to customize the application host for windows when compiling from Linux,
+	# NET 6 is unable to customize the application host for windows when compiling from Linux,
```
Comment (pchote) [inline packaging/functions.sh:156]: Is this still true?

**Review by pchote — COMMENTED** [review]
net5 references in Makefile and packaging/functions.sh will need changing too.

Also the issue with the missing `System.Diagnostics.Debugger` type on macOS (and likely also linux) builds.

**Comment by pchote** [comment]
This can be migrated to the proper release now.

File: OpenRA.Mods.Common/Widgets/Logic/Installation/InstallFromDiscLogic.cs [hunk 2]
```diff
@@ -472,6 +473,11 @@ string FindSourcePath(ModContent.ModSource source, IEnumerable<string> volumes)
 				if (Platform.CurrentPlatform != PlatformType.Windows)
 					return null;
 
+				// We need an extra check for the platform here to silence a warning when the registry is accessed
+				// TODO: Remove this once our platform checks use the same method
+				if (!RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
```
Comment (pchote) [inline OpenRA.Mods.Common/Widgets/Logic/Installation/InstallFromDiscLogic.cs:478]: This could be combined with or replace the check just above.

File: Directory.Build.props [hunk 3]
```diff
@@ -17,10 +17,14 @@
   </PropertyGroup>
 
   <PropertyGroup>
-    <TargetFramework Condition="'$(Mono)' == ''">net5.0</TargetFramework>
+    <TargetFramework Condition="'$(Mono)' == ''">net6.0</TargetFramework>
     <TargetFramework Condition="'$(Mono)' != ''">netstandard2.1</TargetFramework>
   </PropertyGroup>
 
+  <PropertyGroup>
+    <SuppressNETCoreSdkPreviewMessage>true</SuppressNETCoreSdkPreviewMessage>
```
Comment (pchote) [inline Directory.Build.props]: This can probably removed after switching to the real release?

**Comment by pchote** [comment]
It looks like the trimming changes in NET6 have again broken compatibility with our usecase, and there doesn't seem to be any easy workarounds that I could find (its breaking on a bunch of different types to last time).

We might just have to remove the `PublishTrimmed=true` from https://github.com/OpenRA/OpenRA/blob/bleed/packaging/functions.sh#L81 and live with the ~20MB increase in installer size 😞.

File: packaging/functions.sh (hunk 1 above)
Comment (pchote) [reply packaging/functions.sh:156]: > /usr/share/dotnet/sdk/6.0.100/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.targets(550,5): warning NETSDK1074: The application host executable will not be customized because adding resources requires that the build be performed on Windows (excluding Nano Server). [/home/paul2/src/OpenRA/OpenRA.WindowsLauncher/OpenRA.WindowsLauncher.csproj]

Still true.

**Comment by abcdefg30** [comment]
Test installers are at https://github.com/abcdefg30/OpenRA/releases/tag/devtest-20211222.

File: packaging/macos/buildpackage.sh [hunk 4]
```diff
@@ -128,7 +128,7 @@ build_platform() {
 		modify_plist "{MINIMUM_SYSTEM_VERSION}" "10.9" "${TEMPLATE_DIR}/Contents/Info.plist"
 		clang -m64 launcher-mono.m -o "${TEMPLATE_DIR}/Contents/MacOS/Launcher" -framework AppKit -mmacosx-version-min=10.9
 	else
-		modify_plist "{MINIMUM_SYSTEM_VERSION}" "10.13" "${TEMPLATE_DIR}/Contents/Info.plist"
+		modify_plist "{MINIMUM_SYSTEM_VERSION}" "10.14" "${TEMPLATE_DIR}/Contents/Info.plist"
 		clang -m64 launcher.m -o "${TEMPLATE_DIR}/Contents/MacOS/Launcher" -framework AppKit -mmacosx-version-min=10.13
```
Comment (pchote) [inline packaging/macos/buildpackage.sh]: 10.14 here too.

**Review by pchote — COMMENTED** [review]
Makefile still needs updating.

The two macOS and Linux builds run as expected on my machines.

**Comment by reaperrr** [comment]
Needs rebase.

File: packaging/macos/buildpackage.sh (hunk 4 above)
Comment (abcdefg30) [reply packaging/macos/buildpackage.sh]: Fixed, thanks.

**Comment by abcdefg30** [comment]
Rebased and updated the Makefile as well. I went through the csprojs and installed all possible nuget package updates, too.

**Comment by abcdefg30** [comment]
New tag: https://github.com/abcdefg30/OpenRA/releases/tag/devtest-20211229

File: OpenRA.Platforms.Default/OpenRA.Platforms.Default.csproj [hunk 5]
```diff
@@ -3,7 +3,7 @@
     <ProjectReference Include="..\OpenRA.Game\OpenRA.Game.csproj" />
     <PackageReference Include="OpenRA-Freetype6" Version="1.0.4" />
     <PackageReference Include="OpenRA-OpenAL-CS" Version="1.0.16" />
-    <PackageReference Include="OpenRA-SDL2-CS" Version="1.0.28" />
+    <PackageReference Include="OpenRA-SDL2-CS" Version="1.0.29" />
```
Comment (pchote) [inline OpenRA.Platforms.Default/OpenRA.Platforms.Default.csproj]: From discord: we'll want to either revert this or update the package further.

File: OpenRA.WindowsLauncher/OpenRA.WindowsLauncher.csproj [hunk 6]
```diff
@@ -21,7 +21,7 @@
     </AssemblyAttribute>
   </ItemGroup>
   <ItemGroup>
-    <PackageReference Include="OpenRA-SDL2-CS" Version="1.0.28" />
+    <PackageReference Include="OpenRA-SDL2-CS" Version="1.0.29" />
```
Comment (pchote) [inline OpenRA.WindowsLauncher/OpenRA.WindowsLauncher.csproj]: Likewise here.

**Comment by pchote** [comment]
There's one last `net5` that needs to be changed: https://github.com/OpenRA/OpenRA/blob/bf218ecdac8cff033ea9431584cad75d6d8271a4/packaging/functions.sh#L14

**Comment by abcdefg30** [comment]
> There's one last `net5` that needs to be changed:

Fixed, sorry for the sloppiness. I also went back to `1.0.28` for SDL for now.

**Comment by pchote** [comment]
Aaah sorry (sloppyness on my part for only searching `net5` earlier): https://github.com/OpenRA/OpenRA/blob/bleed/INSTALL.md and https://github.com/OpenRA/OpenRA/blob/bleed/configure-system-libraries.sh#L33 need changes too.

**Comment by abcdefg30** [comment]
Updated.

**Review by pchote — DISMISSED** [review]

File: OpenRA.Mods.Common/OpenRA.Mods.Common.csproj [hunk 7]
```diff
@@ -1,9 +1,12 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup Condition="'$(Mono)' != ''">
+    <DefineConstants>MONO</DefineConstants>
+  </PropertyGroup>
```
Comment (penev92) [inline OpenRA.Mods.Common/OpenRA.Mods.Common.csproj]: Why is this here all of a sudden?

**Review by penev92 — COMMENTED** [review]
To be honest I can't be sure whether we're missing anything here or not, but the present changes look OK and it compiles & runs on Windows with no issues both with VS and VSCode.
I want to clear up a few things though:
 - [x] What's with the new `MONO` constant in `OpenRA.Mods.Common`?
 - [ ] AFAIK we want to update the SDL package too and IMO it's best to do it here.
 - [x] Fun fact: the [Pfim](https://nickbabcock.github.io/Pfim/) NuGet package actually released a new version yesterday, so may as well update that?
 - [x] Can you provide binaries (installers) with the latest changes for testing please?

File: OpenRA.Mods.Common/OpenRA.Mods.Common.csproj (hunk 7 above)
Comment (abcdefg30) [reply OpenRA.Mods.Common/OpenRA.Mods.Common.csproj]: This variable is used the disable/enable certain paths on mono, but was never defined, and nobody noticed that before... EDIT: Maybe I'm wrong after all and this *is* a leftover, I'll tag a test build.

**Comment by abcdefg30** [comment]
> AFAIK we want to update the SDL package too and IMO it's best to do it here.

We need to update our own nuget package for that first, which this PR doesn't need to wait on.

**Comment by abcdefg30** [comment]
Generated new builds at https://github.com/abcdefg30/OpenRA/releases/tag/devtest-20220103.

File: OpenRA.Mods.Common/OpenRA.Mods.Common.csproj (hunk 7 above)
Comment (abcdefg30) [reply OpenRA.Mods.Common/OpenRA.Mods.Common.csproj]: Removed it.

**Comment by Mailaender** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/541a6783894cd3c2d4ff00a8552981ef02b3762a...692d73b0d773c0e17f22068bf6a6462a5bf81518)
